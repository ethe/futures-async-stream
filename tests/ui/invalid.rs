// SPDX-License-Identifier: Apache-2.0 OR MIT

#![feature(generators, proc_macro_hygiene, stmt_expr_attributes)]

mod signature {
    use futures_async_stream::stream;

    #[stream(item = ())]
    const fn constness() {} //~ ERROR async stream must be declared as async

    #[stream(item = ())]
    fn variadic(_: ...) {} //~ ERROR only foreign functions are allowed to be C-variadic

    #[stream(item = ())]
    fn asyncness() {} //~ ERROR async stream must be declared as async

    #[stream(item = ())]
    async fn output() -> i32 {} //~ ERROR async stream must return the unit type

    #[stream(item = ())]
    async fn unit() -> () {} // OK

    #[stream(item = ())]
    const async unsafe extern "C" fn f() {} //~ ERROR async stream may not be const
}

mod attribute {
    use futures_async_stream::{for_await, stream, try_stream};

    #[stream(item = ())]
    async fn stream() {}

    #[stream(item = ())]
    #[stream(item = ())] //~ ERROR duplicate #[stream] attribute
    async fn duplicate_stream_fn() {}

    #[try_stream(ok = (), error = ())]
    #[try_stream(ok = (), error = ())] //~ ERROR duplicate #[try_stream] attribute
    async fn duplicate_try_stream_fn() {}

    #[stream(item = ())]
    #[try_stream(ok = (), error = ())] //~ ERROR may not be used at the same time
    async fn combine_fn() {}

    async fn duplicate_stream_async() {
        let _ = {
            #[stream]
            #[stream] //~ ERROR duplicate #[stream] attribute
            async move {}
        };
    }

    async fn duplicate_try_stream_async() {
        let _ = {
            #[try_stream]
            #[try_stream] //~ ERROR duplicate #[try_stream] attribute
            async move {}
        };
    }

    async fn duplicate_for_await() {
        #[for_await]
        #[for_await] //~ ERROR duplicate #[for_await] attribute
        for () in stream() {}
    }

    async fn combine_async() {
        let _ = {
            #[stream]
            #[try_stream] //~ ERROR may not be used at the same time
            async move {}
        };
    }

    #[stream(item = ())]
    async fn duplicate_stream_async_in_fn() {
        let _ = {
            #[stream]
            #[stream] //~ ERROR duplicate #[stream] attribute
            async move {}
        };
    }

    #[stream(item = ())]
    async fn duplicate_try_stream_async_in_fn() {
        let _ = {
            #[try_stream]
            #[try_stream] //~ ERROR duplicate #[try_stream] attribute
            async move {}
        };
    }

    #[stream(item = ())]
    async fn duplicate_for_await_in_fn() {
        #[for_await]
        #[for_await] //~ ERROR duplicate #[for_await] attribute
        for () in stream() {}
    }

    #[stream(item = ())]
    async fn combine_async_in_fn() {
        let _ = {
            #[stream]
            #[try_stream] //~ ERROR may not be used at the same time
            async move {}
        };
    }
}

mod item {
    use futures_async_stream::stream;

    #[stream(item = ())] //~ ERROR #[stream] attribute may only be used on async functions or async blocks
    mod m {}

    #[stream(item = ())] //~ ERROR #[stream] attribute may only be used on async functions or async blocks
    trait A {}

    #[stream(item = ())] //~ ERROR #[stream] attribute may only be used on async functions or async blocks
    impl A {}
}

fn main() {}
