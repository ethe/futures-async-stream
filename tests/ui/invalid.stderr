error: async stream must be declared as async
 --> $DIR/invalid.rs:7:11
  |
7 |     const fn constness() {} //~ ERROR async stream must be declared as async
  |           ^^

error: async stream must be declared as async
  --> $DIR/invalid.rs:10:5
   |
10 |     fn variadic(_: ...) {} //~ ERROR only foreign functions are allowed to be C-variadic
   |     ^^

error: async stream must be declared as async
  --> $DIR/invalid.rs:13:5
   |
13 |     fn asyncness() {} //~ ERROR async stream must be declared as async
   |     ^^

error: async stream must return the unit type
  --> $DIR/invalid.rs:16:26
   |
16 |     async fn output() -> i32 {} //~ ERROR async stream must return the unit type
   |                          ^^^

error: async stream may not be const
  --> $DIR/invalid.rs:22:5
   |
22 |     const async unsafe extern "C" fn f() {} //~ ERROR async stream may not be const
   |     ^^^^^

error: duplicate #[stream] attribute
  --> $DIR/invalid.rs:32:5
   |
32 |     #[stream(item = ())] //~ ERROR duplicate #[stream] attribute
   |     ^^^^^^^^^^^^^^^^^^^^

error: duplicate #[try_stream] attribute
  --> $DIR/invalid.rs:36:5
   |
36 |     #[try_stream(ok = (), error = ())] //~ ERROR duplicate #[try_stream] attribute
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: #[stream] and #[try_stream] may not be used at the same time
  --> $DIR/invalid.rs:40:5
   |
40 |     #[try_stream(ok = (), error = ())] //~ ERROR may not be used at the same time
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error: duplicate #[stream] attribute
  --> $DIR/invalid.rs:46:13
   |
46 |             #[stream] //~ ERROR duplicate #[stream] attribute
   |             ^^^^^^^^^

error: duplicate #[try_stream] attribute
  --> $DIR/invalid.rs:54:13
   |
54 |             #[try_stream] //~ ERROR duplicate #[try_stream] attribute
   |             ^^^^^^^^^^^^^

error: duplicate #[for_await] attribute
  --> $DIR/invalid.rs:61:9
   |
61 |         #[for_await] //~ ERROR duplicate #[for_await] attribute
   |         ^^^^^^^^^^^^

error: #[stream] and #[try_stream] may not be used at the same time
  --> $DIR/invalid.rs:68:13
   |
68 |             #[try_stream] //~ ERROR may not be used at the same time
   |             ^^^^^^^^^^^^^

error: duplicate #[stream] attribute
  --> $DIR/invalid.rs:77:13
   |
77 |             #[stream] //~ ERROR duplicate #[stream] attribute
   |             ^^^^^^^^^

error: duplicate #[try_stream] attribute
  --> $DIR/invalid.rs:86:13
   |
86 |             #[try_stream] //~ ERROR duplicate #[try_stream] attribute
   |             ^^^^^^^^^^^^^

error: duplicate #[for_await] attribute
  --> $DIR/invalid.rs:94:9
   |
94 |         #[for_await] //~ ERROR duplicate #[for_await] attribute
   |         ^^^^^^^^^^^^

error: #[stream] and #[try_stream] may not be used at the same time
   --> $DIR/invalid.rs:102:13
    |
102 |             #[try_stream] //~ ERROR may not be used at the same time
    |             ^^^^^^^^^^^^^

error: #[stream] attribute may only be used on async functions or async blocks
   --> $DIR/invalid.rs:111:5
    |
111 |     #[stream(item = ())] //~ ERROR #[stream] attribute may only be used on async functions or async blocks
    |     ^^^^^^^^^^^^^^^^^^^^
    |
    = note: this error originates in the attribute macro `stream` (in Nightly builds, run with -Z macro-backtrace for more info)

error: #[stream] attribute may only be used on async functions or async blocks
   --> $DIR/invalid.rs:114:5
    |
114 |     #[stream(item = ())] //~ ERROR #[stream] attribute may only be used on async functions or async blocks
    |     ^^^^^^^^^^^^^^^^^^^^
    |
    = note: this error originates in the attribute macro `stream` (in Nightly builds, run with -Z macro-backtrace for more info)

error: #[stream] attribute may only be used on async functions or async blocks
   --> $DIR/invalid.rs:117:5
    |
117 |     #[stream(item = ())] //~ ERROR #[stream] attribute may only be used on async functions or async blocks
    |     ^^^^^^^^^^^^^^^^^^^^
    |
    = note: this error originates in the attribute macro `stream` (in Nightly builds, run with -Z macro-backtrace for more info)
